Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+    simp [← f.map_zpow, ← ofAdd_zsmul]
+#align zpowers_hom zpowersHom
+
+/-- Additive homomorphisms from `ℕ` are defined by the image of `1`. -/
+def multiplesHom [AddMonoid A] :
+    A ≃
+      (ℕ →+
+        A) where
+  toFun : A → ℕ →+ A :=
+  fun x =>
+    {
+    toFun := fun n => n • x
+    map_zero' := zero_nsmul x
+    map_add' := fun _ _ => add_nsmul _ _ _
+  }
+  invFun f := f 1
+  left_inv := one_nsmul
+  right_inv f := AddMonoidHom.ext_nat <| one_nsmul (f 1)
+#align multiples_hom multiplesHom
+
+/-- Additive homomorphisms from `ℤ` are defined by the image of `1`. -/
+def zmultiplesHom [AddGroup A] :
+    A ≃
+      (ℤ →+
+        A) where
+  toFun x := {
+    toFun := fun n => n • x
+    map_zero' := zero_zsmul x
+    map_add' := fun _ _ => add_zsmul _ _ _
+  }
+  invFun f := f 1
+  left_inv := one_zsmul
+  right_inv f := AddMonoidHom.ext_int <| one_zsmul (f 1)
+#align zmultiples_hom zmultiplesHom
+
+attribute [to_additive multiplesHom] powersHom
+
+attribute [to_additive zmultiplesHom] zpowersHom
+
+variable {M G A}
+
+@[simp]
+theorem powers_hom_apply [Monoid M] (x : M) (n : Multiplicative ℕ) :
+    powersHom M x n = x ^ (Multiplicative.toAdd n):=
+  rfl
+#align powers_hom_apply powers_hom_apply
+
+@[simp]
+theorem powers_hom_symm_apply [Monoid M] (f : Multiplicative ℕ →* M) :
+    (powersHom M).symm f = f (Multiplicative.ofAdd 1) :=
+  rfl
+#align powers_hom_symm_apply powers_hom_symm_apply
+
+@[simp]
+theorem zpowers_hom_apply [Group G] (x : G) (n : Multiplicative ℤ) :
+    zpowersHom G x n = x ^ (Multiplicative.toAdd n) :=
+  rfl
+#align zpowers_hom_apply zpowers_hom_apply
+
+@[simp]
+theorem zpowers_hom_symm_apply [Group G] (f : Multiplicative ℤ →* G) :
+    (zpowersHom G).symm f = f (Multiplicative.ofAdd 1) :=
+  rfl
+#align zpowers_hom_symm_apply zpowers_hom_symm_apply
+
+@[simp]
+theorem multiples_hom_apply [AddMonoid A] (x : A) (n : ℕ) : multiplesHom A x n = n • x :=
+  rfl
+#align multiples_hom_apply multiples_hom_apply
+
+attribute [to_additive multiples_hom_apply] powers_hom_apply
+
+@[simp]
+theorem multiples_hom_symm_apply [AddMonoid A] (f : ℕ →+ A) : (multiplesHom A).symm f = f 1 :=
+  rfl
+#align multiples_hom_symm_apply multiples_hom_symm_apply
+
+attribute [to_additive multiples_hom_symm_apply] powers_hom_symm_apply
+
+@[simp]
+theorem zmultiples_hom_apply [AddGroup A] (x : A) (n : ℤ) : zmultiplesHom A x n = n • x :=
+  rfl
+#align zmultiples_hom_apply zmultiples_hom_apply
+
+attribute [to_additive zmultiples_hom_apply] zpowers_hom_apply
+
+@[simp]
+theorem zmultiples_hom_symm_apply [AddGroup A] (f : ℤ →+ A) : (zmultiplesHom A).symm f = f 1 :=
+  rfl
+#align zmultiples_hom_symm_apply zmultiples_hom_symm_apply
+
+attribute [to_additive zmultiples_hom_symm_apply] zpowers_hom_symm_apply
+
+-- TODO use to_additive in the rest of this file
+theorem MonoidHom.apply_mnat [Monoid M] (f : Multiplicative ℕ →* M) (n : Multiplicative ℕ) :
+    f n = f (Multiplicative.ofAdd 1) ^ (Multiplicative.toAdd n) := by
```

Comment:
I think the reason `n.toAdd` is not working is that the coercion to function is not kicking in before Lean 4 gives up, BTW